File: src/log.ts
// Minimal logging helpers. All wrapper chatter goes to stderr so it never
// pollutes the stdout of the git command being wrapped.

const useColor = process.stderr.isTTY && process.env.NO_COLOR === undefined;

const dim = (s: string) => (useColor ? `\x1b[2m${s}\x1b[0m` : s);
const yellow = (s: string) => (useColor ? `\x1b[33m${s}\x1b[0m` : s);
const red = (s: string) => (useColor ? `\x1b[31m${s}\x1b[0m` : s);

/** Low-key informational line (e.g. "switched to account X"). */
export function notice(message: string): void {
  process.stderr.write(dim(`gw: ${message}`) + "\n");
}

export function warn(message: string): void {
  process.stderr.write(yellow(`gw: warning: ${message}`) + "\n");
}

export function error(message: string): void {
  process.stderr.write(red(`gw: error: ${message}`) + "\n");
}

File: src/log.rs
use std::env;
use std::io::IsTerminal;

// All wrapper chatter goes to stderr so it never pollutes the stdout of the
// git command being wrapped.

fn use_color() -> bool {
    std::io::stderr().is_terminal() && env::var_os("NO_COLOR").is_none()
}

fn paint(code: &str, text: &str) -> String {
    if use_color() {
        format!("\x1b[{code}m{text}\x1b[0m")
    } else {
        text.to_string()
    }
}

/// Low-key informational line (e.g. "switched to account X").
pub fn notice(message: &str) {
    eprintln!("{}", paint("2", &format!("gw: {message}")));
}

pub fn warn(message: &str) {
    eprintln!("{}", paint("33", &format!("gw: warning: {message}")));
}

pub fn error(message: &str) {
    eprintln!("{}", paint("31", &format!("gw: error: {message}")));
}
